javatests/arcs/core/data/expression/TypeCheckerTest.kt

```diff
@@ -0,0 +1,300 @@
+    checkAssignableTo(BigInt("1").asExpr(), NumberType)
+
+    checkAssignableTo(1L.asExpr(), LongType)
+    checkAssignableTo(1L.asExpr(), BigIntType)
+    checkAssignableTo(1L.asExpr(), DoubleType)
+    checkAssignableTo(1L.asExpr(), FloatType)
+    checkAssignableTo(1L.asExpr(), NumberType)
+
+    checkAssignableTo(1.asExpr(), IntType)
+    checkAssignableTo(1.asExpr(), BigIntType)
+    checkAssignableTo(1.asExpr(), LongType)
+    checkAssignableTo(1.asExpr(), DoubleType)
+    checkAssignableTo(1.asExpr(), FloatType)
+    checkAssignableTo(1.asExpr(), NumberType)
+
+    checkAssignableTo(1.toShort().asExpr(), ShortType)
+    checkAssignableTo(1.toShort().asExpr(), IntType)
+    checkAssignableTo(1.toShort().asExpr(), BigIntType)
+    checkAssignableTo(1.toShort().asExpr(), LongType)
+    checkAssignableTo(1.toShort().asExpr(), FloatType)
+    checkAssignableTo(1.toShort().asExpr(), DoubleType)
+    checkAssignableTo(1.toShort().asExpr(), NumberType)
+
+    checkAssignableTo(1.toByte().asExpr(), ByteType)
+    checkAssignableTo(1.toByte().asExpr(), ShortType)
+    checkAssignableTo(1.toByte().asExpr(), IntType)
+    checkAssignableTo(1.toByte().asExpr(), BigIntType)
+    checkAssignableTo(1.toByte().asExpr(), LongType)
+    checkAssignableTo(1.toByte().asExpr(), FloatType)
+    checkAssignableTo(1.toByte().asExpr(), DoubleType)
+    checkAssignableTo(1.toByte().asExpr(), NumberType)
+    checkAssignableTo("foo".asExpr(), TextType)
+    checkAssignableTo(true.asExpr(), BooleanType)
+  }
+
+  @Test
+  fun check_unary_ops() {
+    checkAssignableTo(-(1.asExpr()), IntType)
+    checkAssignableTo(-(1.0.asExpr()), DoubleType)
+    checkAssignableTo(!(true.asExpr()), BooleanType)
+  }
+
+  @Test
+  fun check_binary_ops() {
+    checkAssignableTo(1.asExpr() + 2.asExpr(), IntType)
+    checkAssignableTo(1.0.asExpr() + 2.asExpr(), DoubleType)
+
+    checkAssignableTo(1.asExpr() - 2.asExpr(), IntType)
+    checkAssignableTo(1.0.asExpr() - 2.asExpr(), DoubleType)
+
+    checkAssignableTo(1.asExpr() * 2.asExpr(), IntType)
+    checkAssignableTo(1.0.asExpr() * 2.asExpr(), DoubleType)
+
+    checkAssignableTo(1.asExpr() / 2.asExpr(), IntType)
+    checkAssignableTo(1.0.asExpr() / 2.asExpr(), DoubleType)
+
+    checkAssignableTo(1.asExpr() eq 2.asExpr(), BooleanType)
+    checkAssignableTo(1.asExpr() neq 2.asExpr(), BooleanType)
+    checkAssignableTo(1.asExpr() lt 2.asExpr(), BooleanType)
+    checkAssignableTo(1.asExpr() lte 2.asExpr(), BooleanType)
+    checkAssignableTo(1.asExpr() gt 2.asExpr(), BooleanType)
+    checkAssignableTo(1.asExpr() gte 2.asExpr(), BooleanType)
+    checkAssignableTo(1.asExpr() eq 2.asExpr(), BooleanType)
+    checkAssignableTo(1.asExpr() neq 2.asExpr(), BooleanType)
+
+    checkAssignableTo(true.asExpr() and true.asExpr(), BooleanType)
+    checkAssignableTo(true.asExpr() or true.asExpr(), BooleanType)
+
+    checkAssignableTo(1.asExpr() ifNull "hello".asExpr(), UnionType(NumberType, TextType))
```

Comment:
Are there other cases besides `ifNull` that will create union types? I'd like to see more tests for those.

Reply:
ifNull creates union types, and the union() function creates union types, as does first(). Any input handles with union types or nullable types will also create union types, since that is how we represent nullable.